sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/ParDo.java

```diff
@@ -930,6 +983,10 @@ public void populateDisplayData(Builder builder) {
     public PCollectionTuple expand(PCollection<? extends InputT> input) {
       // SplittableDoFn should be forbidden on the runner-side.
       validateWindowType(input, fn);
+
+      // Use coder registry to determine coders for all StateSpec defined in the fn signature.
+      inferStateCodersIfNeeded(fn, input.getPipeline().getCoderRegistry(), input.getCoder());
```

Comment:
@kennknowles - Calling `input.getCoder()` here will throw an exception if the user did not set a coder or registered a default coder in coder registry.
I was thinking if I should do this inside of the inference logic instead and catch this exception. However, our previous discussion about forcing coders to be set regardless of runner makes me think we want this to fail these cases?